### ISSUE TITLE
Is there a way to reduce memory cost?

### ISSUE THREAD

**Issue by xuewyang** [pytorch/pytorch#22139]
I have codes for machine translation. In the decoder output, 
```python
output = F.log_softmax(F.linear(features, self.embed_tokens.weight), dim=-1)
```
Usually, this will give an output tensor of dimension Batch x Sequence_length x dimension, e.g., 50 x 100 x512. Then I use 
`loss = -output.gather(dim=-1, index=target)` to get the final loss.
However, in my case, output is of dimension Batch x Sequence_length x K x dimension, e.g., 50 x 100 x 40 x 512, which is a four dimensional tensor. Thus, the memory cost is huge. Is there a way to reduce this memory cost? Thank you.

**Comment by VitalyFedyunin** [comment]
Please use https://discuss.pytorch.org for questions. Please kindly feel encouraged to reopen this issue, if this does not apply.
